Kernel/Devices/VMWareSVGA/VMWareSVGADevice.cpp

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright (c) 2021, the Serenity OS Developers
+ * Copyright (c) 2021, Alexander Richards <electrodeyt@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+
+#include "VMWareSVGADevice.h"
+#include <AK/Singleton.h>
+#include <Kernel/PCI/Access.h>
+#include <Kernel/VM/AnonymousVMObject.h>
+#include <Kernel/VM/MemoryManager.h>
+#include <LibC/errno_numbers.h>
+#include <LibC/sys/ioctl_numbers.h>
+
+namespace Kernel {
+
+static AK::Singleton<VMWareSVGADevice> s_the;
+
+void VMWareSVGADevice::initialize()
+{
+    s_the.ensure_instance();
+}
+
+VMWareSVGADevice& VMWareSVGADevice::the()
+{
+    return *s_the;
+}
+
+// We put a fake IRQ into the IRQ Handler, as we will change it anyway
+// TODO: this is a really nasty hack that only works as the IRQ is registered on the first call to enable_irq()
+// This should probably be passed in by the init_stage2
+VMWareSVGADevice::VMWareSVGADevice()
+    : BlockDevice(29, 0)
+    , IRQHandler(0)
+{
+    dbgln("VMWareSVGA: init");
+    // Find the VMWare graphics
+    static const PCI::ID vmware_id = { 0x15AD, 0x0405 };
+    PCI::enumerate([&](const PCI::Address& address, PCI::ID id) {
+        if (id == vmware_id) {
+            // Allocate some sheeet
+            m_io_base = PCI::get_BAR0(address) & 0xfffffff0;
+            m_framebuffer_address_physical = PhysicalAddress(page_base_of(PCI::get_BAR1(address)));
+            m_framebuffer_max_size = PCI::get_BAR_space_size(address, 1);
+            m_fifo_mapping = MM.allocate_kernel_region(PhysicalAddress(page_base_of(PCI::get_BAR2(address) & 0xfffffff0)), page_round_up(PCI::get_BAR_space_size(address, 2)), "VMWareSVGA Driver FIFO", Region::Access::Read | Region::Access::Write, Region::Cacheable::No);
+            m_fifo_addr = m_fifo_mapping->vaddr();
+            m_interrupt_line = PCI::get_interrupt_line(address);
+            change_irq_number(m_interrupt_line);
+            dbgln("VMWareSVGA: pci address: {}, io_base: {:x}, framebuffer phys: {:x}", address, m_io_base, m_framebuffer_address_physical.as_ptr());
+            // Clear framebuffer memory with arbitrary value
+            OwnPtr<Region> test_map = MM.allocate_kernel_region(m_framebuffer_address_physical, page_round_up(PCI::get_BAR_space_size(address, 1)), "VMWareSVGA FB Test", Region::Access::Write, Region::Cacheable::No);
+            // memset(test_map->vaddr().as_ptr(), 0xCC, PCI::get_BAR_space_size(address, 1));
+            test_map->unmap();
```

Comment:
is this supposed to still be here?

Reply:
Was there to test, but clearing the framebuffer might actually be a good idea (as it extends beyond the "main" part, and doubles as a DMA area I believe)